### PR TITLE
Change Python's `ObjectID` to `ObjectRef`

### PR DESCRIPTION
**Pull request by raulchen**
## Why are these changes needed?

This PR changes Python's `ObjectID` to `ObjectRef`, due to the following reasons:
1) `ObjectID` is no longer an ID, it is also responsible for other functionalities such as reference counting and asyncio integration.
2) Similar to `ActorHandle`, we should expose high-level handles/references in the API, instead of low-level IDs.
3) Consistent concepts with other language front-ends. 

## PR summary for easier code review

This PR includes the following changes:

- Renamed the following things in all `py/pyx/pxi/pxd/rst` files.
  - `ObjectID` -> `ObjectRef`, except `CObjectID`.
  - `object_id` -> `object_ref`, except `c_object_id` and `c_outer_object_id`. 
  - `obj_id` -> `obj_ref`.
  - `objectid` -> `object_ref`. 
  - `Object/object id/ID` -> `Object/object ref`.
- Moved `ObjectRef` to `object_ref.pxi` file.
- For backward compatibility, added a type alias `ObjectID` to `ObjectRef`. 
- Added `test_object_id_backward_compatibility` unit test.

The following changes aren't included in this PR, because they will change code logic:

- Make `ObjectRef` not inherit from `BaseID`.
- Refine methods in `ObjectRef`. 

## Related issue number

#7291 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: ci/long_running_tests/workloads/many_tasks_serialized_ids.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # This workload stresses distributed reference counting by passing and
-# returning serialized ObjectIDs.
+# returning serialized ObjectRefs.
```
Comment (richardliaw) [inline ci/long_running_tests/workloads/many_tasks_serialized_ids.py:2]: Please make sure to change all of the documentation (ray/doc/source/*.rst) too

Comment (raulchen) [reply ci/long_running_tests/workloads/many_tasks_serialized_ids.py:2]: Sure. Will ping you again when I've changed everything.

Comment (raulchen) [reply ci/long_running_tests/workloads/many_tasks_serialized_ids.py:2]: All python/cython/rst files should have been fixed

File: doc/source/memory-management.rst [hunk 2]
```diff
@@ -3,43 +3,43 @@ Memory Management
 
 This page describes how memory management works in Ray and how you can set memory quotas to ensure memory-intensive applications run predictably and reliably.
 
-ObjectID Reference Counting
+ObjectRef Reference Counting
 ---------------------------
```
Comment (ericl) [inline doc/source/memory-management.rst:7]:
```suggestion
----------------------------
```

**Review by ericl — APPROVED** [review]
lgtm if tests pass

**Review by edoakes — APPROVED** [review]
LGTM!

File: doc/source/memory-management.rst [hunk 3]
```diff
@@ -3,43 +3,43 @@ Memory Management
...
+The ``ray memory`` command can be used to help track down what ``ObjectRef`` references are in scope and may be causing an ``ObjectStoreFullError``.
 
-Running ``ray memory`` from the command line while a Ray application is running will give you a dump of all of the ``ObjectID`` references that are currently held by the driver, actors, and tasks in the cluster.
+Running ``ray memory`` from the command line while a Ray application is running will give you a dump of all of the ``ObjectRef`` references that are currently held by the driver, actors, and tasks in the cluster.
 
 .. code-block::
 
   -----------------------------------------------------------------------------------------------------
-  Object ID                                Reference Type       Object Size   Reference Creation Site
+  Object ref                                Reference Type       Object Size   Reference Creation Site
```
Comment (kfstorm) [inline doc/source/memory-management.rst:42]:
```suggestion
  Object Ref                                Reference Type       Object Size   Reference Creation Site
```

File: doc/source/ray-dashboard.rst [hunk 4]
```diff
@@ -319,7 +319,7 @@ Memory
 
 **Type**: Type of a process. It is either a driver or worker.
 
-**Object ID**: Object ID of a Ray object.
+**Object ref**: Object ref of a Ray object.
```
Comment (kfstorm) [inline doc/source/ray-dashboard.rst:322]:
```suggestion
**Object Ref**: Object ref of a Ray object.
```

File: doc/source/memory-management.rst (hunk 3 above)
Comment (kfstorm) [reply doc/source/memory-management.rst:42]: Same for others

File: doc/source/ray-dashboard.rst (hunk 4 above)
Comment (kfstorm) [reply doc/source/ray-dashboard.rst:322]: Same for others

**Comment by raulchen** [comment]
@edoakes @ericl Do you know why the [ubuntu-jenkins-doc CI](https://github.com/ray-project/ray/runs/854131310?check_suite_focus=true) would fail with the following error? It seems that I need to mock something? 
```
Traceback (most recent call last):
  File "/ray/doc/examples/plot_pong_example.py", line 216, in <module>
    ray.init()
  File "/opt/conda/lib/python3.6/site-packages/ray/worker.py", line 800, in init
    job_id=job_id)
  File "/opt/conda/lib/python3.6/site-packages/ray/worker.py", line 1295, in connect
    ray.internal.free([temporary_object_id])
  File "/opt/conda/lib/python3.6/site-packages/ray/internal/internal_api.py", line 59, in free
    if isinstance(object_refs, ray.ObjectRef):
AttributeError: module 'ray' has no attribute 'ObjectRef'
```

File: python/ray/includes/libcoreworker.pxd [hunk 5]
```diff
@@ -157,7 +157,7 @@ cdef extern from "ray/core_worker/core_worker.h" nogil:
                        c_vector[shared_ptr[CRayObject]] *results)
         CRayStatus Contains(const CObjectID &object_id, c_bool *has_object)
         CRayStatus Wait(const c_vector[CObjectID] &object_ids, int num_objects,
-                        int64_t timeout_ms, c_vector[c_bool] *results)
+                        int64_t timeout_ms, c_vector[c_bool] *results,)
```
Comment (kfstorm) [inline python/ray/includes/libcoreworker.pxd:160]: Maybe revert?

**Comment by raulchen** [comment]
Other jenkins CI also has this error: `AttributeError: module 'ray' has no attribute 'ObjectRef'`.  @rkooo567 Do you know what I missed for jenkins?

Other GitHub CI is even weirder. This seems to be a git issue. I don't know how it's related with this change.
`fatal: Invalid symmetric difference expression 42f8f16c041e01f2066b47204987149d4ac64c2b...a85c81bb7773738f913743f3fa148731c680ac8f`

File: python/ray/includes/libcoreworker.pxd (hunk 5 above)
Comment (raulchen) [reply python/ray/includes/libcoreworker.pxd]: thanks, all fixed.

**Comment by raulchen** [comment]
@simon-mo Could you help take a look at this CI failure? https://github.com/ray-project/ray/runs/854131310?check_suite_focus=true

It seems that the `ubuntu-jenkins-xxx` CI jobs don't use the latest code from this PR.

**Comment by simon-mo** [comment]
If the Amplab jenkins passes, we can ignore the github action Ubuntu-jenkins-* for now. You are correct, they don’t build from source due to disk space issue.

**Comment by raulchen** [comment]
@simon-mo thanks!

File: python/ray/tests/test_reference_counting.py [hunk 6]
```diff
@@ -35,7 +35,7 @@ def _fill_object_store_and_get(oid, succeed=True, object_MiB=40,
         ray.put(np.zeros(object_MiB * 1024 * 1024, dtype=np.uint8))
 
     if type(oid) is bytes:
-        oid = ray.ObjectID(oid)
+        oid = ray.ObjectRef(oid)
```
Comment (kfstorm) [inline python/ray/tests/test_reference_counting.py:38]: Replace `oid` with `oref` or `object_ref`.

Comment (raulchen) [reply python/ray/tests/test_reference_counting.py]: fixed

**Pull request merged**